classes/Subscription.php

```diff
@@ -98,33 +100,74 @@ private function update_subscription($form_data) {
+                    if ($coupon->percent_off == 100) {
+                        $old_price = 0;
+                    } elseif ($coupon->percent_off == 50) {
+                        $old_price /= 2;
+                    }
+                }
+            }
         }
-        \Stripe\Subscription::update($this->stripe->id, $args);
 
-        # Attempt immediate payment on the upgrade
-        try {
-            $invoice = \Stripe\Invoice::create([
-                'customer' => $this->stripe->customer,
-                'subscription' => $this->stripe,
-                'tax_rates': [STRIPE_TAX_RATE],
-            ]);
-            $invoice->finalizeInvoice();
-            $invoice->pay();
-        } catch (\Stripe\Exception\InvalidRequestException $e) {
-            # No invoice created if nothing to pay
-        } catch (\Stripe\Exception\CardException $e) {
-            # A source may still require 3DS... Stripe will have sent an email :-/
+       if ($old_price >= $new_price) {
+            if ($this->stripe->schedule) {
+                \Stripe\SubscriptionSchedule::release($this->stripe->schedule);
+            }
+            $schedule = \Stripe\SubscriptionSchedule::create(['from_subscription' => $this->stripe->id]);
+            $phases = [
+                [
+                    'items' => [['price' => $schedule->phases[0]->items[0]->price]],
+                    'start_date' => $schedule->phases[0]->start_date,
+                    'end_date' => $schedule->phases[0]->end_date,
+                    'proration_behavior' => 'none',
+                    'default_tax_rates' => [STRIPE_TAX_RATE],
+                ],
+                [
+                    'items' => [['price' => $form_data['plan']]],
+                    'iterations' => 1,
+                    'metadata' => $form_data['metadata'],
+                    'proration_behavior' => 'none',
+                    'default_tax_rates' => [STRIPE_TAX_RATE],
+                ],
+            ];
+            if ($schedule->phases[0]->discounts && $schedule->phases[0]->discounts[0]->coupon) {
+                $phases[0]['discounts'] = [['coupon' => $schedule->phases[0]->discounts[0]->coupon]];
+            }
+            if ($form_data['coupon']) {
+                $phases[1]['coupon'] = $form_data['coupon'];
+            }
+            \Stripe\SubscriptionSchedule::update($schedule->id, ['phases' => $phases]);
+        }
+
+        if ($old_price < $new_price) {
+            $args = [
+                'payment_behavior' => 'allow_incomplete',
+                'plan' => $form_data['plan'],
+                'metadata' => $form_data['metadata'],
+                'cancel_at_period_end' => false, # Needed in Stripe 2018-02-28
+                'proration_behavior': 'always_invoice',
+            ];
+            if ($form_data['coupon']) {
+                $args['coupon'] = $form_data['coupon'];
+            } elseif ($this->stripe->discount) {
```

Comment:
I'm assuming this is just saying if there was a discount and there is no longer a discount then clear the discount?

Reply:
Yeah, exactly, I think update only changes the things you provide.